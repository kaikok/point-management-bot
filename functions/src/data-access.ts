import * as admin from "firebase-admin";

let dataAccessFirestore : admin.firestore.Firestore;

export const setFirestore = (firestore: admin.firestore.Firestore) => {
  dataAccessFirestore = firestore;
};

export const getData = async (
  path: string,
  defaultData: admin.firestore.DocumentData):
    Promise<admin.firestore.DocumentData> => {
  let data: admin.firestore.DocumentData = defaultData;
  const docRef = await dataAccessFirestore.doc(path).get();
  if (docRef !== undefined) {
    data = docRef.data() || defaultData;
  }
  return data;
};

export const addDoc = async (
  path: string,
  data: admin.firestore.DocumentData):
    Promise<admin.firestore.DocumentData> => {
  const collectionRef = await dataAccessFirestore.collection(path);
  if (collectionRef === undefined) {
    throw new Error("Collection not found");
  }
  return await collectionRef.add(data);
};

export const setDoc = async (
  path: string,
  docName: string,
  data: admin.firestore.DocumentData):
    Promise<admin.firestore.DocumentData> => {
  const collectionRef = await dataAccessFirestore.collection(path);
  if (collectionRef === undefined) {
    throw new Error("Collection not found");
  }
  return await collectionRef.doc(docName).set(data);
};

export const updateDoc = async (
  path: string,
  data: {[x: string]: unknown;}):
    Promise<admin.firestore.DocumentData> => {
  const docRef = dataAccessFirestore.doc(path);
  return await docRef.update(data);
};

export const deleteDoc = async (
  path: string):
    Promise<admin.firestore.DocumentData> => {
  const docRef = dataAccessFirestore.doc(path);
  return await docRef.delete();
};
