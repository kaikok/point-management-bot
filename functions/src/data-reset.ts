import * as admin from "firebase-admin";


export const dbReset = async (
  firestore: admin.firestore.Firestore) => {
  await firestore.recursiveDelete(firestore.collection("participants"));
  await firestore.recursiveDelete(firestore.collection("items"));
  await firestore.recursiveDelete(firestore.collection("users"));
};
