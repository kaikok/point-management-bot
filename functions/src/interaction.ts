/* eslint-disable require-jsdoc */
import {Context, Telegraf} from "telegraf";
import {Update, ReplyKeyboardRemove}
  from "telegraf/typings/core/types/typegram";
import {createMachine, fromPromise} from "xstate";

export function setupStateMachine(
  bot: Telegraf<Context<Update>>, removeKeyboard: ReplyKeyboardRemove) {
  const machine = createMachine(
    {
      context: ({input}) => ({
        userId: input.userId,
        conversationContext: "",
      }),
      id: "interact",
      states: {
        conversation: {
          initial: "awaitCommand",
          states: {
            awaitCommand: {
              on: {
                eventCommand: {
                  target: "returnQueryTypeKeyboard",
                },
              },
            },
            returnQueryTypeKeyboard: {
              invoke: {
                input: {},
                src: "SendEventQueryTypeKboard",
                onDone: [
                  {
                    target: "awaitEventQueryType",
                    actions: {
                      type: "CheckDone",
                    },
                  },
                ],
                onError: [
                  {
                    target: "awaitCommand",
                    actions: {
                      type: "CheckError",
                    },
                  },
                ],
              },
            },
            awaitEventQueryType: {
              on: {
                receivedInput: {
                  target: "removeKeyboard",
                  guard: "isQueryToday",
                },
              },
            },
            removeKeyboard: {
              invoke: {
                input: {},
                src: "SendEmptyKeyboard",
                onDone: [
                  {
                    target: "awaitCommand",
                  },
                ],
                onError: [
                  {
                    target: "awaitCommand",
                  },
                ],
              },
            },
          },
        },
        hydration: {
          initial: "persisted",
          states: {
            persisted: {
              on: {
                restore: {
                  target: "restored",
                },
              },
            },
            restored: {},
          },
        },
      },
      type: "parallel",
      types: {
        events: {} as
            | { type: "eventCommand" }
            | { type: "receivedInput" }
            | { type: "restore" },
      },
    },
    {
      actions: {
        CheckDone: function({context, event}, params) {
          console.log("Send keyboard done");
        },
        CheckError: function({context, event}, params) {
          console.log("Send keyboard error");
        },
      },
      actors: {
        SendEventQueryTypeKboard: fromPromise(
          async ({
            input,
          }: {
              input: {
                userId: number;
              };
            }) => {
            const builtinKeyboard = {
              resize_keyboard: true,
              one_time_keyboard: true,
              keyboard: [["today"], ["yesterday"], ["ddmmyyyy"]],
            };
            return await bot.telegram.sendMessage(
              input.userId,
              "Events from which day?",
              {reply_markup: builtinKeyboard},
            );
          },
        ),
        SendEmptyKeyboard: fromPromise(
          async ({
            input,
          }: {
              input: {
                userId: number;
              };
            }) => {
            return await bot.telegram.sendMessage(
              input.userId,
              "Back to initial state",
              {reply_markup: removeKeyboard},
            );
          },
        ),
      },
      guards: {
        isQueryToday: function({context, event}, params) {
          console.log("is it today?");
          console.log(event.input);
          return event.input === "today";
        },
      },
      delays: {},
    },
  );
  return machine;
}

module.exports = {
  setupStateMachine: setupStateMachine,
};
