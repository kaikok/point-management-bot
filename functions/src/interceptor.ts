/* eslint-disable require-jsdoc */
import * as functions from "firebase-functions";
import type {Context, Middleware} from "telegraf";
import {Opts, Telegram}
  from "telegraf/typings/core/types/typegram";
import ApiClient from "telegraf/typings/core/network/client";

export function middlewareInterceptor(): Middleware<Context> {
  const middleware: Middleware<Context> = async (ctx, next) => {
    const oldCallApi = ctx.telegram.callApi.bind(ctx.telegram);
    const newCallApi: typeof ctx.telegram.callApi =
      async function newCallApi(
        this: typeof ctx.telegram, method, payload, {signal} = {}) {
        functions.logger.info(
          "ctx-reply-interceptor-request",
          method, payload, signal, {structuredData: true});
        const response = await oldCallApi(method, payload, {signal});
        functions.logger.info(
          "ctx-reply-interceptor-response",
          response, {structuredData: true});
        return response;
      };
    ctx.telegram.callApi = newCallApi.bind(ctx.telegram);
    return next();
  };
  return middleware;
}

export function outgoingInterceptor(oldCallApi: <M extends keyof Telegram>(
  method: M,
  payload: Opts<M>,
  {signal}: ApiClient.CallApiOptions) =>
    Promise<ReturnType<Telegram[M]>>):<M extends keyof Telegram>(
  method: M,
  payload: Opts<M>,
  {signal}: ApiClient.CallApiOptions) => Promise<ReturnType<Telegram[M]>> {
  async function newCallApi<M extends keyof Telegram>(
    this: Telegram,
    method: M,
    payload: Opts<M>,
    {signal} : ApiClient.CallApiOptions = {}):
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      Promise<ReturnType<(...args: any) => any>> {
    functions.logger.info(
      "bot-sendmessage-interceptor-request",
      method, payload, signal, {structuredData: true});
    const response = await oldCallApi(
      method, payload, {signal}) as ReturnType<Telegram[M]>;
    functions.logger.info(
      "bot-sendmessage-interceptor-response",
      response, {structuredData: true});
    return response;
  }
  return newCallApi;
}

module.exports = {
  middlewareInterceptor: middlewareInterceptor,
  outgoingInterceptor: outgoingInterceptor,
};
