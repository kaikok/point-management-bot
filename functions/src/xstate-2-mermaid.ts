const eventLabelTransformers = [
  {
    active: true,
    pattern: /xstate\.done\.actor\.[0-9].*\.([^.]+)$/,
    replacement: "$1.OnDone",
  },
  {
    active: true,
    pattern: /xstate\.error\.actor\.[0-9].*\.([^.]+)$/,
    replacement: "$1.OnError",
  },
  {
    active: true,
    pattern: /xstate\.done.*\.([^.]+)$/,
    replacement: "$1.OnDone",
  },
  {
    active: true,
    pattern: /xstate\.error.*\.([^.]+)$/,
    replacement: "$1.OnError",
  },
];

export const transformEventLabel = (label: string) => {
  let transformed = label;
  eventLabelTransformers.forEach((transformer) => {
    if (transformer.active) {
      const replaced = label.replace(
        transformer.pattern, transformer.replacement);
      if (replaced !== label) {
        transformed = replaced;
      }
    }
  });
  return transformed;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const getInitialState = (stateMachineNode: any) => {
  return stateMachineNode.states[stateMachineNode.config.initial];
};

export const recurseXStateNode = (
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  stateMachineNode: any, rootStateId: string, identation: number) => {
  let mermaidString = "";
  const {id, type, transitions, parent, states} = stateMachineNode;

  if (rootStateId === id) {
    mermaidString = mermaidString.concat("stateDiagram-v2\n");
  }

  mermaidString =
    mermaidString.concat(
      "  ".repeat(identation) +
      `${id} : ${stateMachineNode.key}\n`);

  if (type === "compound") {
    mermaidString =
      mermaidString.concat("  ".repeat(identation) +
        `state ${id} {\n`);
    mermaidString =
      mermaidString.concat("  ".repeat(identation + 1) +
        "[*] --> " +
        `${getInitialState(stateMachineNode).id}\n`);
    if (states) {
      const stateName = Object.keys(states);
      for (const idx in stateName) {
        if (Object.prototype.hasOwnProperty.call(
          stateName, idx)) {
          mermaidString =
            mermaidString.concat(
              recurseXStateNode(states[stateName[idx]],
                rootStateId,
                identation + 1));
        }
      }
    }
    mermaidString =
      mermaidString.concat("  ".repeat(identation) + "}\n");
  }

  if (type === "parallel") {
    mermaidString =
      mermaidString.concat("  ".repeat(identation) +
        `state ${id} {\n`);
    if (states) {
      const stateName = Object.keys(states);
      let first = true;
      for (const idx in stateName) {
        if (Object.prototype.hasOwnProperty.call(
          stateName, idx)) {
          if (first) {
            first = false;
          } else {
            mermaidString = mermaidString.concat(
              "  ".repeat(identation + 1) +
              "--\n");
          }
          mermaidString =
            mermaidString.concat(
              recurseXStateNode(states[stateName[idx]],
                rootStateId,
                identation + 1));
        }
      }
    }
    mermaidString =
      mermaidString.concat("  ".repeat(identation) + "}\n");
  }

  for (const [eventName, transition] of transitions) {
    const targets = transition[0].target;
    const guard = transition[0].guard;

    for (const idx in targets) {
      if (Object.prototype.hasOwnProperty.call(targets, idx)) {
        const srcParentId = parent.id;
        const targetParentId = targets[idx].parent.id;
        if (srcParentId === targetParentId) {
          mermaidString =
            mermaidString.concat(
              "  ".repeat(identation) +
              `${id} --> ${targets[idx].id} :` +
              ` ${transformEventLabel(eventName)}` +
              `${guard ? "/" + guard : ""}\n`);
        } else {
          const siblingStates = targets[idx].parent.states;
          if (siblingStates) {
            const stateName = Object.keys(siblingStates);
            for (const stateIdx in stateName) {
              if (Object.prototype.hasOwnProperty.call(
                stateName, stateIdx)) {
                if (targets[idx].id !== siblingStates[stateName[stateIdx]].id) {
                  mermaidString =
                    mermaidString.concat(
                      "  ".repeat(identation) +
                      `${siblingStates[stateName[stateIdx]].id} -->` +
                      ` ${targets[idx].id} :` +
                      ` ${transformEventLabel(eventName)}` +
                      `${guard ? "/" + guard : ""}\n`);
                }
              }
            }
          }
        }
      }
    }
  }

  if (type === "final") {
    mermaidString =
      mermaidString.concat(
        "  ".repeat(identation) +
        `${id} --> [*]\n`);
  }

  return mermaidString;
};

export const generateMermaid = (
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  stateMachine: any) => {
  return recurseXStateNode(stateMachine.root, stateMachine.root.config.id, 0);
};
