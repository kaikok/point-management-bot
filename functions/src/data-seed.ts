import * as admin from "firebase-admin";
import {setDoc} from "./data-access";
import {Timestamp} from "firebase-admin/firestore";


export const dbSeed = async (
  firestore: admin.firestore.Firestore) => {
  await setDoc(firestore, "participants", "playerA",
    {
      total: 0,
    });
  await setDoc(firestore, "participants/playerA/events", "sampleEvent",
    {
      name: "Test Event",
      timestamp: Timestamp.now(),
    });
  await setDoc(firestore, "participants", "playerB",
    {
      total: 0,
    });
  await setDoc(firestore, "participants/playerB/events", "sampleEvent",
    {
      name: "Test Event",
      timestamp: Timestamp.now(),
    });
  await setDoc(firestore, "items", "demerit",
    {
      descriptions: [
        "not so good behavior",
        "bad behavior",
      ],
      values: [
        -1,
        -2,
      ],
    });
  await setDoc(firestore, "items", "redemption",
    {
      descriptions: [
        "Big prize",
        "Small prize",
        "Small trip",
      ],
      values: [
        -2,
        -1,
        -3,
      ],
    });
  await setDoc(firestore, "items", "score",
    {
      descriptions: [
        "run an errand",
        "do a good deed",
      ],
      values: [
        1,
        2,
      ],
    });
  await setDoc(firestore, "users", "88888888",
    {
      context: "{}",
      role: "admin",
      state: "idle",
      valid: true,
    });
};
