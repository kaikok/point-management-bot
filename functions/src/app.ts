import * as functions from "firebase-functions";
import * as admin from "firebase-admin";
import express from "express";
import * as bodyParser from "body-parser";
import morganBody from "morgan-body";
import "dotenv/config";

import {Telegraf, Context} from "telegraf";
import {KeyboardButton, Message, ReplyKeyboardRemove, Update}
  from "telegraf/typings/core/types/typegram";
import {addDoc, getData, updateDoc} from "./data-access";
import {dbReset} from "./data-reset";
import {dbSeed} from "./data-seed";
import {middlewareInterceptor, outgoingInterceptor}
  from "./interceptor";


if (admin.apps.length === 0) {
  admin.initializeApp();
}

const firestore = admin.firestore();
const app = express();
app.use(bodyParser.json());
app.use("/telegram", bodyParser.json());
morganBody(
  app,
  {
    maxBodyLength: 30000,
    noColors: true,
    prettify: false,
    includeNewLine: false,
    logRequestBody: true,
    logResponseBody: true,
  });

let telegramStatus = "Bot is not loaded";

const removeKeyboard : ReplyKeyboardRemove= {
  "remove_keyboard": true,
  "selective": true,
};

const cmdDocNameLookup : {[key: string]: string} = {
  "score": "score",
  "demerit": "demerit",
  "redeem": "redemption",
};

const getUser = async (
  firestore: admin.firestore.Firestore,
  userId: number): Promise<admin.firestore.DocumentData> => {
  const user = await getData(
    firestore,
    `users/${userId}`,
    {valid: false, role: "none"});
  return user;
};

const isAdmin = (
  user: admin.firestore.DocumentData): boolean => {
  if (user.role === "admin") return true;
  return false;
};

const handlePointsUpdate = async (
  ctx: Context<{
    message: Update.New & Update.NonChannel & Message.TextMessage;
    update_id: number;}>,
  commandName: string,
  participants: Array<string>): Promise<void> => {
  const user = await getUser(firestore, ctx.from.id);
  if (!user.valid) return;
  if (!isAdmin(user)) return;

  const tokens = ctx.message.text.split(" ");

  if (tokens.length < 2) {
    const builtinKeyboard = {
      "resize_keyboard": true,
      "one_time_keyboard": true,
      "keyboard": new Array<Array<KeyboardButton>>(),
    };
    const numOfParticipants = participants.length;
    for (let idx = 0; idx < numOfParticipants; idx++) {
      builtinKeyboard["keyboard"].
        push([`/${commandName} ${participants[idx]}`]);
    }

    functions.logger.info(ctx, {structuredData: true});
    await ctx.reply(`Who do you want to ${commandName}?`,
      {reply_markup: builtinKeyboard});
  } else if (tokens.length === 2 ) {
    const name = tokens[1];
    const aggregate =
      await getData(firestore,
        `items/${cmdDocNameLookup[commandName]}`,
        {descriptions: [], values: []});

    const builtinKeyboard = {
      "resize_keyboard": true,
      "one_time_keyboard": true,
      "keyboard": new Array<Array<KeyboardButton>>(),
    };
    const length = aggregate.descriptions.length;
    for (let idx = 0; idx < length; idx++) {
      builtinKeyboard.keyboard.push(
        [`/${commandName} ${name} ${idx} ${aggregate.descriptions[idx]}`]);
    }

    await ctx.reply("What item?", {reply_markup: builtinKeyboard});
  } else {
    const name = tokens[1];
    const itemIdx = tokens[2];
    const idx = ctx.message.text.search(name);
    const itemName = ctx.message.text.substring(
      idx + name.length + itemIdx.length + 2);

    const aggregate =
      await getData(firestore, `items/${cmdDocNameLookup[commandName]}`,
        {descriptions: [], values: []});

    await addDoc(firestore, `participants/${name}/events`, {
      type: `${cmdDocNameLookup[commandName]}`,
      name: itemName,
      item_id: itemIdx,
      timestamp: admin.firestore.Timestamp.now()});
    await updateDoc(firestore, `participants/${name}`, {
      total: admin.firestore.FieldValue.
        increment(aggregate.values[parseInt(itemIdx)])});
    await ctx.reply(`Done! ${itemName}`, {reply_markup: removeKeyboard});
  }
};

const handleItemsUpdate = async (
  ctx: Context<{
    message: Update.New & Update.NonChannel & Message.TextMessage;
    update_id: number;}>,
  itemClasses: Array<string>): Promise<void> => {
  const user = await getUser(firestore, ctx.from.id);
  if (!user.valid) return;
  if (!isAdmin(user)) return;

  const tokens = ctx.message.text.split(" ");

  if (tokens.length < 2) {
    const builtinKeyboard = {
      "resize_keyboard": true,
      "one_time_keyboard": true,
      "keyboard": new Array<Array<KeyboardButton>>(),
    };
    itemClasses.forEach((itemClass) => {
      builtinKeyboard["keyboard"].push([`/items ${itemClass}`]);
    });

    functions.logger.info(ctx, {structuredData: true});
    await ctx.reply("Which item to query?",
      {reply_markup: builtinKeyboard});
  } else if (tokens.length === 2 ) {
    const itemClass = tokens[1];
    const itemList =
      await getData(firestore, `items/${itemClass}`,
        {descriptions: [], values: []});

    const builtinKeyboard = {
      "resize_keyboard": true,
      "one_time_keyboard": true,
      "keyboard": [
        [`/items ${itemClass} Add`],
        [`/items ${itemClass} Update`],
        [`/items ${itemClass} Cancel`],
      ],
    };

    let reply = "Item : Value\n";
    const length = itemList.descriptions.length;
    for (let idx = 0; idx < length; idx++) {
      reply = reply.concat(
        `${itemList.descriptions[idx]} : ${itemList.values[idx]}\n`);
    }

    await ctx.reply(reply, {reply_markup: builtinKeyboard});
  } else if (tokens.length === 3 ) {
    const itemClass = tokens[1];
    const action = tokens[2];
    if (action === "Add") {
      await updateDoc(
        firestore, `users/${ctx.from.id}`,
        {
          state: "add item",
          context: `{"itemClass":"${itemClass}",`+
          "\"currentField\":\"description\"}"});
      await ctx.reply(
        "What is the description for the new item?",
        {reply_markup: removeKeyboard});
    } else if (action === "Update") {
      const itemList =
        await getData(firestore, `items/${itemClass}`,
          {descriptions: [], values: []});

      const builtinKeyboard = {
        "resize_keyboard": true,
        "one_time_keyboard": true,
        "keyboard": new Array<Array<KeyboardButton>>(),
      };

      const length = itemList.descriptions.length;
      for (let idx = 0; idx < length; idx++) {
        builtinKeyboard.keyboard.push(
          [`/items ${itemClass} Update ${idx} ${itemList.descriptions[idx]}`]);
      }

      await ctx.reply("Which item to update?", {reply_markup: builtinKeyboard});
    } else {
      await ctx.reply("cancelled", {reply_markup: removeKeyboard});
    }
  } else {
    const itemClass = tokens[1];
    const action = tokens[2];
    const itemIdx = tokens[3];

    if (action === "Add") {
      await ctx.reply("cancelled", {reply_markup: removeKeyboard});
    } else if (action === "Update") {
      await updateDoc(
        firestore, `users/${ctx.from.id}`,
        {
          state: "update item",
          context: `{"itemClass":"${itemClass}","itemIdx":${itemIdx}}`});
      await ctx.reply("What is the new value?", {reply_markup: removeKeyboard});
    } else {
      await updateDoc(
        firestore, `users/${ctx.from.id}`,
        {state: "idle", context: "{}"});
      await ctx.reply("cancelled", {reply_markup: removeKeyboard});
    }
  }
};

if (!process.env.YOURTOKEN) {
  functions.logger.info("YOURTOKEN is not defined.", {structuredData: true});
  telegramStatus = "YOURTOKEN is not defined.";
} else {
  const bot = new Telegraf(process.env.YOURTOKEN, {
    telegram: {webhookReply: true},
  });

  const oldCallApi = bot.telegram.callApi.bind(bot.telegram);
  const newCallApi = outgoingInterceptor(oldCallApi);
  bot.telegram.callApi = newCallApi.bind(bot.telegram);

  const middleware = middlewareInterceptor();
  bot.use(middleware);

  bot.hears("hi", (ctx) => {
    const message = `Hey there, it is ${(new Date()).toLocaleString()} now.`;
    ctx.reply(message, {reply_markup: removeKeyboard});
    functions.logger.info(message, {structuredData: true});
    functions.logger.info(
      `${ctx.message.from.username}: ${ctx.from.id}: ${ctx.message.chat.id}`,
      {structuredData: true});
  });

  bot.command("items", async (ctx) => {
    await handleItemsUpdate(ctx,
      ["score", "demerit", "redemption"]);
  });

  bot.command("points", async (ctx) => {
    const playerAPoints =
      await getData(firestore, "participants/playerA", {total: 0});
    const playerBPoints =
      await getData(firestore, "participants/playerB", {total: 0});

    let reply = "Participant : Score\n";
    reply = reply.concat(`playerA : ${playerAPoints.total}\n`);
    reply = reply.concat(`playerB : ${playerBPoints.total}\n`);
    await ctx.reply(reply, {reply_markup: removeKeyboard});
    return;
  });

  bot.command("score", async (ctx) => {
    const commandName = "score";
    const participants = ["playerA", "playerB"];
    await handlePointsUpdate(ctx, commandName, participants);
  });

  bot.command("demerit", async (ctx) => {
    const commandName = "demerit";
    const participants = ["playerA", "playerB"];
    await handlePointsUpdate(ctx, commandName, participants);
  });

  bot.command("redeem", async (ctx) => {
    const commandName = "redeem";
    const participants = ["playerA", "playerB"];
    await handlePointsUpdate(ctx, commandName, participants);
  });

  bot.on("text", async (ctx) => {
    const userId = ctx.from.id;
    const user = await getData(
      firestore,
      `users/${userId}`,
      {
        valid: false,
      });

    if (!user.valid) return;

    if (isAdmin(user)) {
      switch (user.state) {
      case "update item": {
        const context = JSON.parse(user.context);
        const itemList = await getData(
          firestore, `items/${context.itemClass}`,
          {descriptions: [], values: []});
        if (itemList.descriptions.length === 0) {
          await ctx.reply(`No ${context.itemClass} items found.`,
            {reply_markup: removeKeyboard});
          return;
        }
        const itemDescription = itemList.descriptions[context.itemIdx];
        const newValue = parseInt(ctx.message.text);
        const newValuesList = itemList.values;
        newValuesList[context.itemIdx] = newValue;
        await updateDoc(firestore,
          `items/${context.itemClass}`, {
            values: newValuesList, descriptions: itemList.descriptions});
        await updateDoc(
          firestore,
          `users/${userId}`,
          {state: "idle", context: "{}"});
        await ctx.reply(
          `Updated ${context.itemClass} item,`+
          `"${itemDescription}" with value "${newValue}"`,
          {reply_markup: removeKeyboard});
        return;
      }
      case "add item": {
        const context = JSON.parse(user.context);
        if (context.currentField === "description") {
          const description = ctx.message.text;
          const newContext = `{"itemClass":"${context.itemClass}",`+
            `"currentField":"value","description":"${description}"}`;

          await updateDoc(
            firestore,
            `users/${userId}`,
            {state: "add item", context: newContext});
          await ctx.reply(
            `Ok description is "${description}", please provide the value.`,
            {reply_markup: removeKeyboard});
        } else if (context.currentField === "value") {
          const itemList = await getData(
            firestore, `items/${context.itemClass}`,
            {descriptions: [], values: []});
          const itemValue = parseInt(ctx.message.text);
          const itemDescription = context.description;

          itemList.descriptions.push(itemDescription);
          itemList.values.push(itemValue);
          await updateDoc(firestore,
            `items/${context.itemClass}`,
            {
              descriptions: itemList.descriptions,
              values: itemList.values,
            });
          await updateDoc(
            firestore, `users/${ctx.from.id}`,
            {state: "idle", context: "{}"});
          await ctx.reply(
            `Added new ${context.itemClass},`+
            ` "${itemDescription}" of value "${itemValue}"`,
            {reply_markup: removeKeyboard});
        }
        return;
      }
      default: {
        return;
      }
      }
    }
  });

  app.use("/telegram", bot.webhookCallback("/",
    {secretToken: process.env.API_TOKEN}));
  telegramStatus = "Bot is loaded.";
}

app.get("/hello", async (req, res) => {
  res.send("Hello, Firebase!");
  functions.logger.log(`hello, ${telegramStatus}`);
  return;
});

if (process.env.FUNCTIONS_EMULATOR) {
  app.get("/dbreset", async (req, res) => {
    dbReset(firestore);
    res.send("Done!");
  });

  app.get("/dbseed", async (req, res) => {
    dbSeed(firestore);
    res.send("Done!");
  });
}

export default app;
