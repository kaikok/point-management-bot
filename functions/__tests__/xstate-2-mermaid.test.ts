jest.mock("telegraf");
import {Telegraf} from "telegraf";
import {setupStateMachine} from "../src/interaction";
import {ReplyKeyboardRemove} from "telegraf/typings/core/types/typegram";
import {generateMermaid} from "../src/xstate-2-mermaid";
import {createMachine} from "xstate";

const removeKeyboard : ReplyKeyboardRemove= {
  "remove_keyboard": true,
  "selective": true,
};

describe("Generate mermaid diagram", () => {
  test.only("get machine", async () => {
    const bot = new Telegraf("", {
      telegram: {webhookReply: true},
    });
    const eventViewingMachine = setupStateMachine(bot, removeKeyboard);
    expect(generateMermaid(eventViewingMachine)).toMatchSnapshot();
  });

  test.only("simple machine", async () => {
    const simple = createMachine(
      {
        id: "simple",
        initial: "lightsOff",
        states: {
          lightsOn: {
            on: {
              toggle: {
                target: "lightsOff",
              },
            },
          },
          lightsOff: {
            on: {
              toggle: {
                target: "lightsOn",
              },
            },
          }
        }
      },
      {
        actions: {},
        actors: {},
        guards: {},
        delays: {},
      });
    console.log(generateMermaid(simple));
    expect(generateMermaid(simple)).toMatchSnapshot();
  });

  test.only("parallel machine", async () => {
    const parallel = createMachine({
      id: "coffee",
      initial: "preparing",
      states: {
        preparing: {
          states: {
            grindBeans: {
              initial: "grindingBeans",
              states: {
                grindingBeans: {
                  on: {
                    BEANS_GROUND: {
                      target: "beansGround",
                    },
                  },
                },
                beansGround: {
                  type: "final",
                },
              },
            },
            boilWater: {
              initial: "boilingWater",
              states: {
                boilingWater: {
                  on: {
                    WATER_BOILED: {
                      target: "waterBoiled",
                    },
                  },
                },
                waterBoiled: {
                  type: "final",
                },
              },
            },
          },
          type: "parallel",
          onDone: {
            target: "makingCoffee",
          },
        },
        makingCoffee: {},
      },
    });

    expect(generateMermaid(parallel)).toMatchSnapshot();
  });
});