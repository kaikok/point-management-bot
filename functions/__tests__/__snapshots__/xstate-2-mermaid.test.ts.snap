// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generate mermaid diagram get machine 1`] = `
"stateDiagram-v2
interact : interact
state interact {
  interact.conversation : conversation
  state interact.conversation {
    [*] --> interact.conversation.awaitCommand
    interact.conversation.awaitCommand : awaitCommand
    interact.conversation.awaitCommand --> interact.conversation.returnQueryTypeKeyboard : eventCommand
    interact.conversation.returnQueryTypeKeyboard : returnQueryTypeKeyboard
    interact.conversation.returnQueryTypeKeyboard --> interact.conversation.awaitEventQueryType : returnQueryTypeKeyboard.OnDone
    interact.hydration.restored --> interact.hydration.persisted : returnQueryTypeKeyboard.OnDone
    interact.conversation.returnQueryTypeKeyboard --> interact.conversation.awaitCommand : returnQueryTypeKeyboard.OnError
    interact.hydration.restored --> interact.hydration.persisted : returnQueryTypeKeyboard.OnError
    interact.conversation.awaitEventQueryType : awaitEventQueryType
    interact.conversation.awaitEventQueryType --> interact.conversation.removeKeyboard : receivedInput/isQueryToday
    interact.conversation.removeKeyboard : removeKeyboard
    interact.conversation.removeKeyboard --> interact.conversation.awaitCommand : removeKeyboard.OnDone
    interact.hydration.restored --> interact.hydration.persisted : removeKeyboard.OnDone
    interact.conversation.removeKeyboard --> interact.conversation.awaitCommand : removeKeyboard.OnError
    interact.hydration.restored --> interact.hydration.persisted : removeKeyboard.OnError
  }
  --
  interact.hydration : hydration
  state interact.hydration {
    [*] --> interact.hydration.persisted
    interact.hydration.persisted : persisted
    interact.hydration.persisted --> interact.hydration.restored : restore
    interact.hydration.restored : restored
  }
}
"
`;

exports[`Generate mermaid diagram parallel machine 1`] = `
"stateDiagram-v2
coffee : coffee
state coffee {
  [*] --> coffee.preparing
  coffee.preparing : preparing
  state coffee.preparing {
    coffee.preparing.grindBeans : grindBeans
    state coffee.preparing.grindBeans {
      [*] --> coffee.preparing.grindBeans.grindingBeans
      coffee.preparing.grindBeans.grindingBeans : grindingBeans
      coffee.preparing.grindBeans.grindingBeans --> coffee.preparing.grindBeans.beansGround : BEANS_GROUND
      coffee.preparing.grindBeans.beansGround : beansGround
      coffee.preparing.grindBeans.beansGround --> [*]
    }
    --
    coffee.preparing.boilWater : boilWater
    state coffee.preparing.boilWater {
      [*] --> coffee.preparing.boilWater.boilingWater
      coffee.preparing.boilWater.boilingWater : boilingWater
      coffee.preparing.boilWater.boilingWater --> coffee.preparing.boilWater.waterBoiled : WATER_BOILED
      coffee.preparing.boilWater.waterBoiled : waterBoiled
      coffee.preparing.boilWater.waterBoiled --> [*]
    }
  }
  coffee.preparing --> coffee.makingCoffee : preparing.OnDone
  coffee.makingCoffee : makingCoffee
}
"
`;

exports[`Generate mermaid diagram simple machine 1`] = `
"stateDiagram-v2
simple : simple
state simple {
  [*] --> simple.lightsOff
  simple.lightsOn : lightsOn
  simple.lightsOn --> simple.lightsOff : toggle
  simple.lightsOff : lightsOff
  simple.lightsOff --> simple.lightsOn : toggle
}
"
`;
